Software-Construction/Factory-And-Abstract-Factory/Graphics/Screen.java

```diff
@@ -0,0 +1,177 @@
+
+		if(noOfObjects == 0){
+			throw new AssertionError("List is empty! Cannot sort");
+		}
+
+		for(int i = 0; i < noOfObjects; i++){
+			for(int j = 0; j < noOfObjects - 1 - i; j++){
+				if(sortedListByArea.get(j).getArea() > sortedListByArea.get(j + 1).getArea()){
+					Shapes temp = sortedListByArea.get(j);
+					sortedListByArea.set(j, sortedListByArea.get(j + 1));
+					sortedListByArea.set(j + 1, temp);
+				}
+			}
+		}
+		return sortedListByArea;
+	}
+
+	/**
+	 * Method to sort the shapes on the basis of their perimeter.
+	 * @return : The sorted list of the shape.
+	 */
+	public List<Shapes> sortByPerimeter(){
+		List<Shapes> sortedListByPerimeter = listOfShapes;
+		int noOfObjects = sortedListByPerimeter.size();
+
+		if(noOfObjects == 0){
+			throw new AssertionError("List is empty! Cannot sort");
+		}
+
+		for(int i = 0; i < noOfObjects; i++){
+			for(int j = 0; j < noOfObjects - 1 - i; j++){
+				if (sortedListByPerimeter.get(j).getPerimeter() > sortedListByPerimeter.get(j + 1).getPerimeter()) {
+					Shapes temp = sortedListByPerimeter.get(j);
+					sortedListByPerimeter.set(j, sortedListByPerimeter.get(j + 1));
+					sortedListByPerimeter.set(j + 1, temp);
+				}
+			}
+		}
+		return sortedListByPerimeter;
+	}
+
+	/**
+	 * Method to sort the shapes based on their distance from the origin of the screen.
+	 * @return : The sorted list of the shape.
+	 */
+	public List<Shapes> sortByOriginDistance(){
+		List<Shapes> sortedListByOriginDistance = listOfShapes;
+		int noOfObjects = sortedListByOriginDistance.size();
+
+		if(noOfObjects == 0){
+			throw new AssertionError("List is empty! Cannot sort");
+		}
+
+		for(int i = 0; i < noOfObjects; i++){
+			for(int j = 0; j < noOfObjects - 1 - i; j++){
+				if (sortedListByOriginDistance.get(j).getOriginDistance() > sortedListByOriginDistance.get(j + 1).getOriginDistance()) {
+					Shapes temp = sortedListByOriginDistance.get(j);
+					sortedListByOriginDistance.set(j, sortedListByOriginDistance.get(j + 1));
+					sortedListByOriginDistance.set(j + 1, temp);
+				}
+			}
+		}
+		return sortedListByOriginDistance;
+	}
+
+	/**
+	 * Method to sort the shape on the basis of the time on which they were added on the screen.
+	 * @return : The sorted list of the shape.
+	 */
+	public List<Shapes> sortByTimestamp(){
+		if (listOfShapes.size() == 0){
+			throw new AssertionError("List is empty! Cannot sort");
+		}
+		return listOfShapes;
+	}
+
+	/**
+	 * Method to Check if the given point to draw the shape doesn't overlap the existing shape.
+	 * @param point : The point where the shape has to be drawn.
+	 * @return : The list of the shape.
+	 */
+	public List<Shapes> shapesEnclosingPoint(Point point){
+		List<Shapes> listOfShapes = new ArrayList<>();
+
+		if(point == null){
+			throw new AssertionError("Enter a valid point");
+		}
+
+		for(Shapes shape : listOfShapes){
+			if(shape.isPointEnclosed(point)){
+				listOfShapes.add(shape);
+			}
+		}
+		return listOfShapes;

```

Comment:
performing all operations on a single list can lead to data loss. What do u think?

Reply:
Actually it is passing a reference of a newly created list. And it will get appended to the listOfShape defined for screen object.